src/mscorlib/shared/System/Collections/Concurrent/ConcurrentQueue_Segment.cs

```diff
@@ -0,0 +1,333 @@
+                if (diff == 0)
+                {
+                    // We may be racing with other enqueuers.  Try to reserve the slot by incrementing
+                    // the tail.  Once we've done that, no one else will be able to write to this slot,
+                    // and no dequeuer will be able to read from this slot until we've written the new
+                    // sequence number. WARNING: The next few lines are not reliable on a runtime that
+                    // supports thread aborts. If a thread abort were to sneak in after the CompareExchange
+                    // but before the Volatile.Write, other threads will spin trying to access this slot.
+                    // If this implementation is ever used on such a platform, this if block should be
+                    // wrapped in a finally / prepared region.
+                    if (Interlocked.CompareExchange(ref _headAndTail.Tail, currentTail + 1, currentTail) == currentTail)
+                    {
+                        // Successfully reserved the slot.  Note that after the above CompareExchange, other threads
+                        // trying to return will end up spinning until we do the subsequent Write.
+                        _slots[slotsIndex].Item = item;
+                        Volatile.Write(ref _slots[slotsIndex].SequenceNumber, currentTail + 1);
+                        return true;
+                    }
+                }
+                else if (diff < 0)
+                {
+                    // The sequence number was less than what we needed, which means this slot still
+                    // contains a value, i.e. the segment is full.  Technically it's possible that multiple
+                    // dequeuers could have read concurrently, with those getting later slots actually
+                    // finishing first, so there could be spaces after this one that are available, but
+                    // we need to enqueue in order.
+                    return false;
+                }
+
+                // Lost a race. Spin a bit, then try again.
+                spinner.SpinOnce();
+            }
+        }
+
+        /// <summary>Represents a slot in the queue.</summary>
+        [StructLayout(LayoutKind.Auto)]
+        [DebuggerDisplay("Item = {Item}, SequenceNumber = {SequenceNumber}")]
+        internal struct Slot
+        {
+            /// <summary>The item.</summary>
+            public T Item;
+            /// <summary>The sequence number for this slot, used to synchronize between enqueuers and dequeuers.</summary>
+            public int SequenceNumber;
+        }
+
+        /// <summary>Padded head and tail indices, to avoid false sharing between producers and consumers.</summary>
+        [DebuggerDisplay("Head = {Head}, Tail = {Tail}")]
+        [StructLayout(LayoutKind.Explicit, Size = 3*Internal.PaddingHelpers.CACHE_LINE_SIZE)] // padding before/between/after fields
```

Comment:
@jkotas PaddedHeadAndTail is just moved here from ConcurrentQueue. Because I had to move Segment class out of ConcurrentQueue as it's type is need in System.Net.Http: 
https://github.com/dotnet/corefx/blob/b399fe53343295ccf475262552b6d36beafb2b4e/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs#L20